src/GitVersionTask/GitVersionTaskBase.cs

```diff
@@ -1,72 +1,141 @@
-                return true;
+                output = execute(input, logger);
             }
             catch (WarningException errorException)
             {
-                LogWarning(errorException.Message);
-                return true;
+                logger.LogWarning(errorException.Message);
+                output = new TOutput();
             }
             catch (Exception exception)
             {
-                LogError("Error occurred: " + exception);
-                return false;
+                logger.LogError("Error occurred: " + exception);
+                throw;
             }
+            finally
+            {
+                Logger.Reset();
+            }
+
+            return output;
         }
 
-        protected abstract void InnerExecute();
 
-        protected ExecuteCore ExecuteCore { get; }
+        public static ExecuteCore CreateExecuteCore()
+            => new ExecuteCore(new FileSystem());
+
+        private static string GetFileExtension(this String language)
+        {
+            switch (language)
+            {
+                case "C#":
+                    return "cs";
 
-        [Required]
-        public string SolutionDirectory { get; set; }
+                case "F#":
+                    return "fs";
 
-        public bool NoFetch { get; set; }
+                case "VB":
+                    return "vb";
 
-        public void LogDebug(string message)
-        {
-            BuildEngine.LogMessageEvent(new BuildMessageEventArgs(message, string.Empty, "GitVersionTask", MessageImportance.Low));
+                default:
+                    throw new Exception($"Unknown language detected: '{language}'");
+            }
         }
 
-        public void LogWarning(string message)
+        public static FileWriteInfo GetWorkingDirectoryAndFileNameAndExtension(
+            this String intermediateOutputPath,
+            String language,
+            String projectFile,
+            Func<String, String, String> fileNameWithIntermediatePath,
+            Func<String, String, String> fileNameNoIntermediatePath
+            )
         {
-            BuildEngine.LogWarningEvent(new BuildWarningEventArgs(string.Empty, string.Empty, null, 0, 0, 0, 0, message, string.Empty, "GitVersionTask"));
+            var fileExtension = language.GetFileExtension();
+            String workingDirectory, fileName;
+            if (intermediateOutputPath == null)
+            {
+                fileName = fileNameWithIntermediatePath(projectFile, fileExtension);
+                workingDirectory = TempFileTracker.TempPath;
+            }
+            else
+            {
+                workingDirectory = intermediateOutputPath;
+                fileName = fileNameNoIntermediatePath(projectFile, fileExtension);
+            }
+            return new FileWriteInfo(workingDirectory, fileName, fileExtension);
         }
 
-        public void LogInfo(string message)
+        public abstract class AbstractInput
```

Comment:
I think `InputBase` is a better name for `abstract` classes. Also, can you move this out into its own file so it's not nested within another class?

```suggestion
        public abstract class InputBase
```

Reply:
Is having it side-by-side with `GitVersionBase` in the same file OK?